cookbooks/memcached/recipes/install_server.rb

```
@@ -0,0 +1,179 @@
+    variables(
+            :tcp_port         => node[:memcached][:tcp_port],
+            :udp_port         => node[:memcached][:udp_port],
+            :user             => node[:memcached][:user],
+            :connection_limit => node[:memcached][:connection_limit],
+            :memtotal         => node[:memcached][:memtotal],               # calculated option
+            :threads          => node[:memcached][:threads],
+            :ip               => node[:memcached][:ip],
+            :log_level        => node[:memcached][:log_level]
+    )
+    cookbook 'memcached'
+end
+
+service "memcached" do
+    action :restart # restart needed for new settings to apply
+end # had to move the restart out of the template due to start problem after server reboot
+
+log "    Memcached configuration done."
+
+
+## firewall configuration
+log "    Attention: when using a listening public ip make sure the #{node[:memcached][:tcp_port]} port is open in the firewall (Security Group for EC2)."
+log "    Opening port #{node[:memcached][:tcp_port]} in iptables."
+
+sys_firewall "Open memcached port" do
+    port node[:memcached][:tcp_port].to_i # port should be passed as int
+    enable true
+    action :update
+end
+
+
+## checking if memcached actually started
+#   problem: when starting memcached on amazon with a public listening ip the daemon doesn't really start though says so
+#   there is no interface with public ip on amazon thus you can find
+#   "failed to listen on TCP port XXXXX: Cannot assign requested address" in /var/log/memcached.log
+#   therefor must use "any ip" aka 0.0.0.0 to listen externally
+ruby_block "memcached_check" do
+    block do
+        # assigning test_ip
+        if "#{node[:memcached][:ip]}" == "0.0.0.0"
+            test_ip = "#{node[:cloud][:public_ips][0]}" # can't use 0.0.0.0 for TCPSocket
+        else
+            test_ip = "#{node[:memcached][:ip]}"        # localhost, private, public are ok
+        end
```

We need to support binding to any interface that is on a server.  Some private clouds may want to use a "public/2nd" interface.  There are also cross cloud use cases that require binding to the public IP.  This is based on what the services template does (and is in use with customers).

The problem is that an amazon user can specify public and expect to bind to the public interface.  This fails because the public IP interface does not exist on an instance.  In this case setting the ip to the public ip would cause memcache to fail silently.  This test_ip is used to verify that memcache did start and strand the server in the case where it did not.

We can go over this in the morning.